src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,130 @@
+                    + listSize);
+        }
+    }
+
+    private void isIndexInRangeInclusive(int index) {
+        if (!(index >= 0 && index <= listSize)) {
+            throw new ArrayListIndexOutOfBoundsException("Index: "
+                    + index
+                    + " is out of bounds. "
+                    + "Size: "
+                    + listSize);
+        }
+    }
+
+    private void growList() {
+        if (listSize == defaultArray.length) {
+            Object [] extendedArray = new Object[(int) (defaultArray.length * ADD_CAPACITY)];
+            System.arraycopy(defaultArray,
+                    POSITION_ZERO,
+                    extendedArray,
+                    POSITION_ZERO,
+                    defaultArray.length);
+            defaultArray = (T[]) extendedArray;
+        }
+    }
+
+    private void insertElement(Object [] array, Object value, int index) {
+        isIndexInRangeInclusive(index);
+        System.arraycopy(array,
+                index, array,
+                index + 1,
+                defaultArray.length - index - 1);
+        array[index] = value;
+        listSize++;
+    }
+
+    private void removeElement(Object[] array, int index) {
+        System.arraycopy(array, index + 1, array, index, defaultArray.length - index - 1);
+        listSize--;
+    }
+
     @Override
     public void add(T value) {
-
+        growList();
+        defaultArray[listSize++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        growList();
+        insertElement(defaultArray,value,index);
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexInBounds(index);
+        final T returnedElement = defaultArray[index];
+        return returnedElement;
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexInBounds(index);
+        defaultArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexInBounds(index);
+        final T removedElement = defaultArray[index];
+        removeElement(defaultArray,index);
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < listSize; i++) {
+            if (defaultArray[i] == element || element != null && element.equals(defaultArray[i])) {
+                return remove(i);
+            }
+        }
+        throw new NoSuchElementException("Element not found: " + element);
     }
 
     @Override
     public int size() {
-        return 0;
+        return listSize;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return listSize == 0;
+    }
+
+    @Override
+    public String toString() {
+        return Arrays.toString(defaultArray);
```

Comment:
The 'toString' method should not use 'Arrays.toString' directly on 'defaultArray' as it will include null values in the output. Instead, consider using 'Arrays.copyOf' to copy only the valid elements up to 'listSize' into a new array and then using 'Arrays.toString' on that array.